src/redisraft.c

```diff
@@ -370,6 +380,172 @@ static void handleInfoCommand(RedisRaftCtx *rr,
+            if (password_len > 255) {
+                LOG_WARNING("password is too long: limited to 255 characters");
+                goto fail;
+            }
+            strncpy(ret->password, password, password_len);
+            break;
+        }
+    }
+
+    return ret;
+
+fail:
+    RedisModule_Free(ret);
+    return NULL;
+}
+
+static RaftReq * cmdToMigrate(RedisRaftCtx *rr, RedisModuleCtx * ctx, RaftRedisCommand *cmd)
+{
+    RaftReq *req = NULL;
+
+    // 0. assert minimum size here
+    if (cmd->argc < 8) {
+        RedisModule_WrongArity(ctx);
+        return req;
+    }
+
+    // 1. Extract migration info and store it in a dict rr->migrate_info
+    MigrationInfo * migrationInfo = extractMigrationInfo(rr, cmd);
+    if (migrationInfo == NULL) {
+        RedisModule_ReplyWithError(ctx, "ERR check logs");
+        goto exit;
+    }
+
+    // 2. assert single key is empty
+    size_t key_len;
+    RedisModule_StringPtrLen(cmd->argv[3], &key_len);
+    if (key_len != 0) {
+        RedisModule_ReplyWithError(ctx, "ERR RedisRaft doesn't support old style single key migration");
+        goto exit;
+    }
+
+    /* find "keys" for parsing out keys */
+    int idx = MIGRATE_CMD_OPTIONAL_ARG_IDX;
+    for(; idx < cmd->argc; idx++) {
+        size_t str_len;
+        const char * str = RedisModule_StringPtrLen(cmd->argv[idx], &str_len);
+        if (str_len == 4 && !strcasecmp("keys", str)) {
+            break;
+        }
+    }
+
+    /* move idx past "keys" argv */
+    idx++;
+
+    if (idx >= cmd->argc) {
+        RedisModule_ReplyWithError(ctx, "ERR Didn't provide any keys to migrate");
+        goto exit;
+    }
+
+    RedisModuleString **keys = RedisModule_Alloc(sizeof(RedisModuleString *) * (cmd->argc - idx));
+    int num_keys = cmd->argc - idx;
+    for (int i = 0; i < num_keys; i++) {
+        keys[i] = RedisModule_HoldString(ctx, cmd->argv[idx + i]);
+    }
+
+    req = RaftReqInit(ctx, RR_MIGRATE_KEYS);
+
+    /* Overwrite with new data */
+    req->r.migrate_keys.num_keys = num_keys;
+    req->r.migrate_keys.keys = keys;
+    req->r.migrate_keys.keys_serialized = RedisModule_Calloc(num_keys, sizeof(RedisModuleString *));
+    memcpy(req->r.migrate_keys.auth_username, migrationInfo->username, 255);
+    memcpy(req->r.migrate_keys.auth_password, migrationInfo->password, 255);
+
+exit:
+    RedisModule_Free(migrationInfo);
+    return req;
+}
+
+static void handleMigrateCommand(RedisRaftCtx *rr, RedisModuleCtx *ctx, RaftRedisCommand *cmd)
+{
+      if (rr->migrate_req != NULL) {
+        RedisModule_ReplyWithError(ctx, "ERR RedisRaft only supports one concurrent migration currently");
+        return;
+    }
+  
+    if (!raft_is_leader(rr->raft)) {
+        Node *leader = getLeaderNodeOrReply(rr, ctx);
+        if (leader) {
+            redirectCommand(rr, ctx, leader);
+        }
+    }
+  
+    RaftReq * req;
+    if ((req = cmdToMigrate(rr, ctx, cmd)) == NULL) {
+        return;
+    }
```

Comment:
```suggestion
    RaftReq *req = cmdToMigrate(rr, ctx, cmd);
    if (req == NULL) {
        return;
    }
```
I'd just comment for pointer alignment but maybe moving assignment out is easier to read.